sw/device/tests/rv_core_ibex_epmp_test.c

```diff
@@ -0,0 +1,336 @@
+
+  CSR_WRITE(CSR_REG_PMPADDR0, tor_address(start));
+  CSR_WRITE(CSR_REG_PMPADDR1, tor_address(end));
+
+  const uint32_t pmp1cfg = (EPMP_CFG_A_TOR | EPMP_CFG_LRWX) ^ EPMP_CFG_R;
+  CSR_SET_BITS(region_pmpcfg(1), pmp1cfg << region_offset(1));
+}
+
+/**
+ * Sets up the permissions for the test locations.
+ *
+ * See the declaration of `test_locations`
+ * to see the desired permission settings.
+ */
+static void pmp_setup_test_locations(void) {
+  CSR_WRITE(CSR_REG_PMPADDR3, tor_address((uintptr_t)(test_locations + 0)));
+  CSR_WRITE(CSR_REG_PMPADDR4, tor_address((uintptr_t)(test_locations + 1)));
+  CSR_WRITE(CSR_REG_PMPADDR5, tor_address((uintptr_t)(test_locations + 2)));
+  CSR_WRITE(CSR_REG_PMPADDR6, tor_address((uintptr_t)(test_locations + 3)));
+  CSR_WRITE(CSR_REG_PMPADDR7, tor_address((uintptr_t)(test_locations + 4)));
+
+  uint32_t cfg = EPMP_CFG_A_TOR | EPMP_CFG_R;
+  CSR_SET_BITS(region_pmpcfg(4), cfg << region_offset(4));
+  cfg = EPMP_CFG_A_TOR | EPMP_CFG_LRW;
+  CSR_SET_BITS(region_pmpcfg(5), cfg << region_offset(5));
+  cfg = EPMP_CFG_A_TOR | EPMP_CFG_W;
+  CSR_SET_BITS(region_pmpcfg(6), cfg << region_offset(6));
+  cfg = EPMP_CFG_A_TOR | EPMP_CFG_L | EPMP_CFG_X | EPMP_CFG_W;
+  CSR_SET_BITS(region_pmpcfg(7), cfg << region_offset(7));
+}
+
+/**
+ * Sets up the UART connection.
+ */
+static void setup_uart(void) {
+  // Initialise DIF handles
+  CHECK_DIF_OK(dif_pinmux_init(
+      mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR), &pinmux));
+  CHECK_DIF_OK(dif_uart_init(
+      mmio_region_from_addr(TOP_EARLGREY_UART0_BASE_ADDR), &uart0));
+
+  // Initialise UART console.
+  pinmux_testutils_init(&pinmux);
+  CHECK(kUartBaudrate <= UINT32_MAX, "kUartBaudrate must fit in uint32_t");
+  CHECK(kClockFreqPeripheralHz <= UINT32_MAX,
+        "kClockFreqPeripheralHz must fit in uint32_t");
+  CHECK_DIF_OK(dif_uart_configure(
+      &uart0, (dif_uart_config_t){
+                  .baudrate = (uint32_t)kUartBaudrate,
+                  .clk_freq_hz = (uint32_t)kClockFreqPeripheralHz,
+                  .parity_enable = kDifToggleDisabled,
+                  .parity = kDifUartParityEven,
+                  .tx_enable = kDifToggleEnabled,
+                  .rx_enable = kDifToggleEnabled,
+              }));
+  base_uart_stdout(&uart0);
+}
+
+/**
+ * The entry point of the SRAM program.
+ *
+ * *Control flow passed from `sram_start`.*
+ */
+void sram_main(void) {
+  setup_uart();
+
+  // Must set up the Machine Mode Area Correctly
+  // before entering Machine Mode Lockdown.
+  pmp_setup_machine_area();
+
+  LOG_INFO("Enable Machine Mode Lockdown");
+  CSR_SET_BITS(CSR_REG_MSECCFG, EPMP_MSECCFG_MML);
+
+  pmp_setup_user_area();
+  pmp_setup_test_locations();
+
+  LOG_INFO("The PMP Config:");
+  dbg_print_epmp();
+
+  LOG_INFO("Machine Mode Tests");
+  uint32_t load;
+  for (int loc = 0; loc < NUM_LOCATIONS; ++loc) {
+    test_locations[loc] = 42;
+    load = test_locations[loc];
+    ((void (*)(void))(test_locations + loc))();
```

Comment:
What is `load` for? Are you always jumping to address 42?

Reply:
Lines:
136. Tries to write to the location
137. Tries to read from the location
138. Tries to execute the location

Reply:
I see, now I understand the test logic, brilliant.